javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll

```diff
@@ -192,16 +182,41 @@ module DomBasedXss {
    * HTML by a jQuery method.
    */
   predicate isPrefixOfJQueryHtmlString(DataFlow::Node htmlString, DataFlow::Node prefix) {
-    any(JQuery::MethodCall call).interpretsArgumentAsHtml(htmlString) and
-    prefix = htmlString
+    prefix = getAPrefixOfJQuerySelectorString(htmlString)
+  }
+
+  /**
+   * Holds if `prefix` is a prefix of `htmlString`, which may be intepreted as
+   * HTML by a jQuery method.
+   */
+  private DataFlow::Node getAPrefixOfJQuerySelectorString(DataFlow::Node htmlString) {
+    any(JQuery::MethodCall call).interpretsArgumentAsSelector(htmlString) and
+    result = htmlString
     or
-    exists(DataFlow::Node pred | isPrefixOfJQueryHtmlString(htmlString, pred) |
-      prefix = StringConcatenation::getFirstOperand(pred)
+    exists(DataFlow::Node pred | pred = getAPrefixOfJQuerySelectorString(htmlString) |
+      result = StringConcatenation::getFirstOperand(pred)
       or
-      prefix = pred.getAPredecessor()
+      result = pred.getAPredecessor()
     )
   }
 
+  /**
+   * An argument to the jQuery `$` function, which is interpreted as either a selector
+   * or as an HTML string depending on its first character.
+   */
+  class JQuerySelectorSink extends Sink {
```

Comment:
Two things here:

- naming: this is a sink because it could interpret the argument as HTML rather than a benign selector.
- reuse: could we unify this with `AmbiguousHtmlOrSelectorArgument`? https://github.com/github/codeql/blob/3587235b4f26a23467b46ed4bea9be5045d6ad8e/javascript/ql/src/semmle/javascript/security/dataflow/UnsafeJQueryPluginCustomizations.qll#L34-L37

Reply:
Renamed to `JQueryHtmlOrSelector...` and shared some logic with `AmbiguousHtmlOrSelectorArgument`.